### PR TITLE
Change featured tag updates to add/remove activity

### PR DESCRIPTION
**Pull request by noellabo**
When updating a featured tag, instead of delivering update activity for the entire Actor, it delivers add/remove activity for the Hashtag.

### CODE REVIEW

File: app/controllers/api/v1/featured_tags_controller.rb [hunk 1]
```diff
@@ -6,30 +6,23 @@ class Api::V1::FeaturedTagsController < Api::BaseController
...
-    ActivityPub::UpdateDistributionWorker.perform_in(3.minutes, current_account.id)
-    render json: @featured_tag, serializer: REST::FeaturedTagSerializer
+    featured_tag = FeaturedTagService.new.call(current_account, featured_tag_params[:name])
+    render json: featured_tag, serializer: REST::FeaturedTagSerializer
   end
 
   def destroy
-    @featured_tag.destroy!
-    ActivityPub::UpdateDistributionWorker.perform_in(3.minutes, current_account.id)
+    UnfeaturedTagWorker.perform_async(current_account.id, params[:id])
```
Comment (Gargron) [inline app/controllers/api/v1/featured_tags_controller.rb:20]: I think you may still want to confirm the record exists before queueing into Sidekiq.

File: app/serializers/activitypub/add_hashtag_serializer.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class ActivityPub::AddHashtagSerializer < ActivityPub::Serializer
```
Comment (Gargron) [inline app/serializers/activitypub/add_hashtag_serializer.rb:3]: I'm not sure if this serializer has to be separate if the only difference is the type of `proper_object` we should be able to use a single `AddSerializer` for everything, see e.g. `Collection` and `Activity`

File: app/services/featured_tag_service.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class FeaturedTagService < BaseService
```
Comment (Gargron) [inline app/services/featured_tag_service.rb:3]: `CreateFeaturedTagService`?

File: app/services/unfeatured_tag_service.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+class UnfeaturedTagService < BaseService
```
Comment (Gargron) [inline app/services/unfeatured_tag_service.rb:3]: `RemoveFeaturedTagService`?

File: app/workers/activitypub/update_distribution_worker.rb [hunk 5]
```diff
@@ -1,8 +1,6 @@
 # frozen_string_literal: true
 
 class ActivityPub::UpdateDistributionWorker < ActivityPub::RawDistributionWorker
-  sidekiq_options queue: 'push', lock: :until_executed
```
Comment (Gargron) [inline app/workers/activitypub/update_distribution_worker.rb:4]: Why?

File: app/workers/unfeatured_tag_worker.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class UnfeaturedTagWorker
```
Comment (Gargron) [inline app/workers/unfeatured_tag_worker.rb:3]: `RemoveFeaturedTagWorker`?

File: app/serializers/activitypub/remove_hashtag_serializer.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class ActivityPub::RemoveHashtagSerializer < ActivityPub::Serializer
```
Comment (Gargron) [inline app/serializers/activitypub/remove_hashtag_serializer.rb:3]: Same as for `Add`

File: app/workers/activitypub/update_distribution_worker.rb (hunk 5 above)
Comment (noellabo) [reply app/workers/activitypub/update_distribution_worker.rb:4]: This sidekiq_options was added at https://github.com/mastodon/mastodon/pull/19380/commits/00b7b9573acd746a366dd01f0abad88417afcbbc.

With this fix, this worker is no longer needed as it is no longer run in perform_in.
But you might not have to remove.

File: app/controllers/api/v1/featured_tags_controller.rb (hunk 1 above)
Comment (noellabo) [reply app/controllers/api/v1/featured_tags_controller.rb:20]: Undo set_featured_tag brefore_action and check it again.

File: app/services/featured_tag_service.rb (hunk 3 above)
Comment (noellabo) [reply app/services/featured_tag_service.rb:3]: Rename it.

File: app/services/unfeatured_tag_service.rb (hunk 4 above)
Comment (noellabo) [reply app/services/unfeatured_tag_service.rb:3]: Rename it.

File: app/workers/unfeatured_tag_worker.rb (hunk 6 above)
Comment (noellabo) [reply app/workers/unfeatured_tag_worker.rb:3]: Rename it.

File: app/workers/activitypub/update_distribution_worker.rb (hunk 5 above)
Comment (noellabo) [reply app/workers/activitypub/update_distribution_worker.rb]: Undo it.

**Review by Gargron — APPROVED** [review]

**Pull request merged**